src/main/java/graphql/schema/GraphQLAppliedDirectiveArgument.java

```diff
@@ -0,0 +1,252 @@
+ * that have been supplied on a {@link GraphQLDirective}.
+ * <p>
+ * The difference is the 'value' and 'defaultValue' properties.  In a query argument, the 'value' is never in the GraphQLArgument
+ * object but rather in the AST direct or in the query variables map and the 'defaultValue' represents a value to use if both of these are
+ * not present. You can think of them like a descriptor of what shape an argument might have.
+ * <p>
+ * However with directives on SDL elements, the value is specified in AST only and transferred into the GraphQLArgument object and the
+ * 'defaultValue' comes instead from the directive definition elsewhere in the SDL.  You can think of them as 'instances' of arguments, their shape and their
+ * specific value on that directive.
+ */
+@PublicApi
+public class GraphQLAppliedDirectiveArgument implements GraphQLNamedSchemaElement {
+
+    private final String name;
+    private final InputValueWithState value;
+
+    private final Argument definition;
+
+
+    private GraphQLAppliedDirectiveArgument(String name,
+                                            InputValueWithState value,
+                                            Argument definition
+    ) {
+        assertValidName(name);
+        this.name = name;
+        this.value = value;
+        this.definition = definition;
+    }
+
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    public boolean hasSetValue() {
+        return value.isSet();
+    }
+
+    /**
+     * @return an input value with state for an applied directive argument
+     */
+    public @NotNull InputValueWithState getArgumentValue() {
+        return value;
+    }
+
+    /**
+     * This static helper method will give out a java value based on the semantics captured
+     * in the {@link InputValueWithState} from {@link GraphQLAppliedDirectiveArgument#getArgumentValue()}
+     *
+     * Note : You MUST only call this on a {@link GraphQLAppliedDirectiveArgument} that is part of a fully formed schema.  We need
+     * all the types to be resolved in order for this work correctly.
+     *
+     * Note: This method will return null if the value is not set or explicitly set to null.  If you want to know the difference
+     * when "not set" and "set to null" then you can't use this method.  Rather you should use {@link GraphQLAppliedDirectiveArgument#getArgumentValue()}
+     * and use the {@link InputValueWithState#isNotSet()} methods to decide how to handle those values.
+     *
+     * @param argument the fully formed {@link GraphQLAppliedDirectiveArgument}
+     * @param <T>      the type you want it cast as
+     *
+     * @return a value of type T which is the java value of the argument
+     */
+    public static <T> T getArgumentValue(GraphQLInputType argumentType, GraphQLAppliedDirectiveArgument argument) {
+        return getInputValueImpl(argumentType, argument.getArgumentValue());
+    }
+
```

Comment:
some question as above: do we actually have a Description?

Reply:
See above - no